### PR TITLE
New Features in VSCode Plugin: Auto server termination after max idle time and default installation of python + jupyter notebook extensions

### PR DESCRIPTION
**Pull request by JasonZhu1313**
New Features in VSCode Plugin: Auto server termination after max idle time and default installation of python + jupyter notebook extensions

# TL;DR
_Please replace this text with a description of what this PR accomplishes._

This is a followup PR of VSCode Plugin contribution: https://github.com/flyteorg/flytekit/pull/1922 with additional features:
1. Terminate the VSCode server after reaching the maximum idle time when no activity is detected.
2. Enabling the installation of VSCode plugins from VSCode plugin marketplace urls, by default python and jupyter notebook plugins.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

- Termination after max idle time
Replace original hard coded server_up_seconds with max_idle_seconds, check the heartbeat file located at ~/.local/share/code-server/heartbeat ever 60 seconds [source](https://coder.com/docs/code-server/latest/FAQ#what-is-the-heartbeat-file), if the duration if the time till last connection is longer than max idle seconds, terminate the VSCode server. 

- Installation of python + jupyter notebook extensions from urls
Expose plugins_remote_paths argument in vscode annotation, so users could choose plugin urls they'd like to install. By default python and Jupyter notebook plugin will be installed. 

The urls of plugin follows this pattern: 

```
https://${publisher}.gallery.vsassets.io/_apis/public/gallery/publisher/${publisher}/extension/${extension name}/${version}/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage
```

We'll be extracting the ${extension name} from url using regex as the extension name on local disk, the extension will be named with .vsix extension.

Ex. Python plugin url

```
https://ms-python.gallery.vsassets.io/_apis/public/gallery/publisher/ms-python/extension/python/2023.12.0/assetbyname/Microsoft.VisualStudio.Services.VSIXPackage
```
Will be downloaded on local disk as python.vsix for installation.

## Tracking Issue
https://github.com/flyteorg/flytekit/pull/1922

## Tests

The delta (duration in seconds after last activity detected) is refreshed after connected to vscode server. And server is terminated after delta reaches to max idle time.

```
{"asctime": "2023-11-17 08:29:41,806", "name": "flytekit", "levelname": "INFO", "message": "heart beat path exist?: True"}
{"asctime": "2023-11-17 08:29:41,806", "name": "flytekit", "levelname": "INFO", "message": "The latest activity on code server is 238.5610237121582 seconds ago."}
{"asctime": "2023-11-17 08:29:41,806", "name": "flytekit", "levelname": "INFO", "message": "metadata"}
{"asctime": "2023-11-17 08:29:41,806", "name": "flytekit", "levelname": "INFO", "message": "238.5610237121582"}
{"asctime": "2023-11-17 08:29:41,806", "name": "flytekit", "levelname": "INFO", "message": "36000"}
{"asctime": "2023-11-17 08:29:41,806", "name": "flytekit", "levelname": "INFO", "message": "False"}
{"asctime": "2023-11-17 08:30:41,867", "name": "flytekit", "levelname": "INFO", "message": "heart beat path is here: /home/flytekit/.local/share/code-server/heartbeat"}
{"asctime": "2023-11-17 08:30:41,885", "name": "flytekit", "levelname": "INFO", "message": "heart beat path exist?: True"}
{"asctime": "2023-11-17 08:30:41,886", "name": "flytekit", "levelname": "INFO", "message": "The latest activity on code server is 14.362635612487793 seconds ago."}
{"asctime": "2023-11-17 08:30:41,886", "name": "flytekit", "levelname": "INFO", "message": "metadata"}
{"asctime": "2023-11-17 08:30:41,887", "name": "flytekit", "levelname": "INFO", "message": "14.362635612487793"}
{"asctime": "2023-11-17 08:30:41,887", "name": "flytekit", "levelname": "INFO", "message": "36000"}
{"asctime": "2023-11-17 08:30:41,888", "name": "flytekit", "levelname": "INFO", "message": "False"}
{"asctime": "2023-11-17 08:31:41,922", "name": "flytekit", "levelname": "INFO", "message": "heart beat path is here: /home/flytekit/.local/share/code-server/heartbeat"}
{"asctime": "2023-11-17 08:31:41,923", "name": "flytekit", "levelname": "INFO", "message": "heart beat path exist?: True"}
{"asctime": "2023-11-17 08:31:41,923", "name": "flytekit", "levelname": "INFO", "message": "The latest activity on code server is 14.39638066291809 seconds ago."}
{"asctime": "2023-11-17 08:31:41,923", "name": "flytekit", "levelname": "INFO", "message": "metadata"}
{"asctime": "2023-11-17 08:31:41,923", "name": "flytekit", "levelname": "INFO", "message": "14.39638066291809"}
{"asctime": "2023-11-17 08:31:41,923", "name": "flytekit", "levelname": "INFO", "message": "36000"}
{"asctime": "2023-11-17 08:31:41,923", "name": "flytekit", "levelname": "INFO", "message": "False"}
```

Plugins are installed successfully
Screenshot 2023-11-18 at 9 54 40 AM

### CODE REVIEW

File: plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py [hunk 1]
```diff
@@ -92,45 +146,68 @@ def download_vscode(
 
     # Download remote file to local
     code_server_tar_path = download_file(code_server_remote_path, DOWNLOAD_DIR)
+    
+    plugin_paths = []
+    for plugin in plugins_remote_paths:
```
Comment (ByronHsu) [inline plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py:151]: We should use the plugin in Open-VSX instead of the official vscode marketplace due to legal reason. Refer to the [codeserver doc](https://coder.com/docs/code-server/latest/FAQ#why-cant-code-server-use-microsofts-extension-marketplace):
> Because of this, we can't offer any extensions on Microsoft's marketplace. Instead, we use the [Open-VSX extension gallery](https://open-vsx.org/), which is also used by various other forks. It isn't perfect, but its getting better by the day with more and more extensions.

You can find plugins on [OpenVSX](https://open-vsx.org/) . For example:
python plugin: https://open-vsx.org/api/ms-python/python/2023.20.0/file/ms-python.python-2023.20.0.vsix
jupyter plugin: https://open-vsx.org/api/ms-toolsai/jupyter/2023.9.100/file/ms-toolsai.jupyter-2023.9.100.vsix
Note: you can find the exact link by "right click" and "copy link address" in the plugin homepage.

File: plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py [hunk 2]
```diff
@@ -92,45 +146,68 @@ def download_vscode(
...
+        execute_command(f"code-server --install-extension {p}")
 
 
 def vscode(
     _task_function: Optional[Callable] = None,
-    server_up_seconds: Optional[int] = DEFAULT_UP_SECONDS,
+    max_idle_seconds: Optional[int] = MAX_IDLE_SECONDS,
     port: Optional[int] = 8080,
     enable: Optional[bool] = True,
     code_server_remote_path: Optional[str] = DEFAULT_CODE_SERVER_REMOTE_PATH,
```
Comment (ByronHsu) [inline plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py:184]: Im considering to wrap all the url related arguments to a dataclass to provide a cleaner interface since users usually won't change the default url.

Can we do

```python
@dataclass
class FlyinConfig:
  code_server_remote_path: DEFAULT_CODE_SERVER_REMOTE_PATH
  code_server_dir_name: DEFAULT_CODE_SERVER_DIR_NAME
  plugins_remote_paths: [...default values]
``` 

And the interface becomes

```python
def vscode(
  config=FlyinConfig(...)
)
```

Note: I will later rename vscode plugin to flyin plugin. So I call the config `FlyinConfig`

File: plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py [hunk 3]
```diff
@@ -148,26 +225,19 @@ def inner_wrapper(*args, **kwargs):
...
 
-            # 3. Terminates the server after server_up_seconds
-            logger.info(f"{server_up_seconds} seconds passed. Terminating...")
-            if post_execute is not None:
-                post_execute()
-                logger.info("Post execute function executed successfully!")
-            child_process.terminate()
-            child_process.join()
-            sys.exit(0)
+            return
```
Comment (ByronHsu) [inline plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py:240]: nit: we don't need to have return here since the program will end in exit_handler

**Comment by ByronHsu** [comment]
nit you have to resolve the DCO issue by adding sign-off to your commit. Check [details](https://github.com/flyteorg/flytekit/pull/1977/checks?check_run_id=18828627242)

**Comment by JasonZhu1313** [comment]
> 7 expected, 1 pending, and 2 successful checks

File: plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py (hunk 3 above)
Comment (JasonZhu1313) [reply plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py:240]: Done thanks

File: plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py (hunk 2 above)
Comment (JasonZhu1313) [reply plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py:184]: Done thanks

File: plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py (hunk 1 above)
Comment (JasonZhu1313) [reply plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py]: Thanks for suggestion, updated pr

File: plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py [hunk 4]
```diff
@@ -35,13 +44,56 @@ def execute_command(cmd):
...
+        if delta > max_idle_seconds:
+            logger.info(f"Container is idle for more than {max_idle_seconds} seconds. Terminating...")
+            if post_execute is not None:
+                post_execute()
+                logger.info("Post execute function executed successfully!")
+            child_process.terminate()
+            child_process.join()
+            sys.exit()
+
+        time.sleep(60)
```
[the remark this answers is not in the document]
Comment (JasonZhu1313) [reply plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py:99]: Thanks, make sense, done.

File: plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py [hunk 5]
```diff
@@ -35,13 +59,56 @@ def execute_command(cmd):
...
 
-def download_file(url, target_dir="."):
+def exit_handler(
+    max_idle_seconds: int = 180,
+    child_process: multiprocessing.Process = None,
+    post_execute: Optional[Callable] = None
+):
+    """
+    Check the modified time of ~/.local/share/code-server/heartbeat.
+    If it is older than max_idle_second seconds, kill the container.
```
Comment (troychiu) [inline plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py:69]: nit: kill the container -> kill the server

File: plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py [hunk 6]
```diff
@@ -35,13 +59,56 @@ def execute_command(cmd):
...
+            delta = (time.time() - start_time)
+            logger.info(f"Code server has not been connected since {delta} seconds ago.")
+            logger.info(f"Please open the browser to connect to the running server.")
+        else:
+            delta = (time.time() - os.path.getmtime(HEARTBEAT_PATH))
+            logger.info(f"The latest activity on code server is {delta} seconds ago.")
+
+        # If the time till last connection is longer than max idle seconds, terminate the vscode server.
+        if delta > max_idle_seconds:
+            logger.info(f"Container is idle for more than {max_idle_seconds} seconds. Terminating...")
```
Comment (troychiu) [inline plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py:91]: nit: Container -> VSCode server

File: plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py [hunk 7]
```diff
@@ -5,20 +5,44 @@
...
+
+# Default max idle seconds to terminate the vscode server
+HOURS_TO_SECONDS = 60 * 60
+MAX_IDLE_SECONDS = 10 * HOURS_TO_SECONDS  # 10 hours
+
+# Duration to pause the checking of the heartbeat file until the next one
+HEARTBEAT_CHECK_SECONDS = 60
+
+# The path is hardcoded by code-server
+# https://coder.com/docs/code-server/latest/FAQ#what-is-the-heartbeat-file
```
Comment (troychiu) [inline plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py:29]: I think you can also move these to constants.py

**Review by troychiu — CHANGES_REQUESTED** [review]

File: plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ByronHsu) [reply plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py:69]: IMO saying "container" here is more precise because since the purpose of exit_hanlder is to shutdown the container

File: plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py (hunk 7 above)
Comment (JasonZhu1313) [reply plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py:29]: Make sense thanks!

File: plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py [hunk 8]
```diff
@@ -35,13 +59,56 @@ def execute_command(cmd):
     logger.info(f"stderr: {stderr}")
 
 
-def download_file(url, target_dir="."):
+def exit_handler(
+    max_idle_seconds: int = 180,
+    child_process: multiprocessing.Process = None,
```
[the remark this answers is not in the document]
Comment (JasonZhu1313) [reply plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py:64]: Good catch, done

File: plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (JasonZhu1313) [reply plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py:69]: +1, the whole container / resources will be recycled instead of code server. (e.g., you can't ssh to the container again after this process is terminated)

File: plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (JasonZhu1313) [reply plugins/flytekit-vscode/flytekitplugins/vscode/decorator.py]: Addressed thanks

**Comment by ByronHsu** [comment]
Thanks jason for the great work! I have validated the following use cases, and they all worked well

```python
from flytekit import task, workflow
from flytekitplugins.vscode import vscode, VscodeConfig
from flytekitplugins.vscode.constants import DEFAULT_CODE_SERVER_EXTENSIONS

@task(
    container_image="localhost:30000/flytekit-vscode:0.0.2",
    environment={"FLYTE_SDK_LOGGING_LEVEL": "20"}
)
@vscode(

)
def t():
    ...



@task(
    container_image="localhost:30000/flytekit-vscode:0.0.2",
    environment={"FLYTE_SDK_LOGGING_LEVEL": "20"}
)
@vscode(
    max_idle_seconds=10,
)
def t_suicide():
    ...



config_with_vim = VscodeConfig(
    extension_remote_paths=DEFAULT_CODE_SERVER_EXTENSIONS+["https://open-vsx.org/api/vscodevim/vim/1.27.0/file/vscodevim.vim-1.27.0.vsix"]
)

@task(
    container_image="localhost:30000/flytekit-vscode:0.0.2",
    environment={"FLYTE_SDK_LOGGING_LEVEL": "20"}
)
@vscode(
    config=config_with_vim
)
def t_vim():
    ...



@workflow
def wf():
    t()
    t_suicide()
    t_vim()

```

**Comment by JasonZhu1313** [comment]
> Thanks jason for the great work! I have validated the following use cases, and they all worked well
> 
> ```python
> from flytekit import task, workflow
> from flytekitplugins.vscode import vscode, VscodeConfig
> from flytekitplugins.vscode.constants import DEFAULT_CODE_SERVER_EXTENSIONS
> 
> @task(
>     container_image="localhost:30000/flytekit-vscode:0.0.2",
>     environment={"FLYTE_SDK_LOGGING_LEVEL": "20"}
> )
> @vscode(
> 
> )
> def t():
>     ...
> 
> 
> 
> @task(
>     container_image="localhost:30000/flytekit-vscode:0.0.2",
>     environment={"FLYTE_SDK_LOGGING_LEVEL": "20"}
> )
> @vscode(
>     max_idle_seconds=10,
> )
> def t_suicide():
>     ...
> 
> 
> 
> config_with_vim = VscodeConfig(
>     extension_remote_paths=DEFAULT_CODE_SERVER_EXTENSIONS+["https://open-vsx.org/api/vscodevim/vim/1.27.0/file/vscodevim.vim-1.27.0.vsix"]
> )
> 
> @task(
>     container_image="localhost:30000/flytekit-vscode:0.0.2",
>     environment={"FLYTE_SDK_LOGGING_LEVEL": "20"}
> )
> @vscode(
>     config=config_with_vim
> )
> def t_vim():
>     ...
> 
> 
> 
> @workflow
> def wf():
>     t()
>     t_suicide()
>     t_vim()
> ```

Thanks! Nice integrate test covers all cases

File: plugins/flytekit-vscode/README.md [hunk 9]
```diff
@@ -39,3 +39,59 @@ If users want to skip the vscode downloading process at runtime, they have the o
 + RUN tar -xzf /tmp/code-server/code-server-4.18.0-linux-amd64.tar.gz -C /tmp/code-server/
 + ENV PATH="/tmp/code-server/code-server-4.18.0-linux-amd64/bin:${PATH}"
 ```
+
+## More Examples
+
+```python
+from flytekit import task, workflow
```
Comment (pingsutw) [inline plugins/flytekit-vscode/README.md:46]: could we also add this example to flytesnacks?

Comment (ByronHsu) [reply plugins/flytekit-vscode/README.md:46]: sure i have added that to our progress tracker: https://docs.google.com/document/d/1NkjCbYoOqz-b9l-3UxZnUzpefsZiA7SC-mD5_Dp6dMQ/edit#bookmark=id.4fpr1ayjw6gj

**Comment by troychiu** [comment]
LGTM. Thank you @JasonZhu1313 and @ByronHsu!

**Review by troychiu — APPROVED** [review]

File: plugins/flytekit-vscode/README.md (hunk 9 above)
Comment (pingsutw) [reply plugins/flytekit-vscode/README.md:46]: thank you

**Comment by troychiu** [comment]
I am wondering if it's possible or even needed to add tests to exit handler?

**Review by pingsutw — APPROVED** [review]
